src/sage/rings/function_field/drinfeld_modules/finite_drinfeld_module.py

```diff
@@ -231,39 +233,183 @@ def frobenius_charpoly(self, var='X'):
 
         ::
 
-            sage: trace = phi.frobenius_trace()
-            sage: trace
-            (4*z3^2 + 6*z3 + 3)*T + 3*z3^2 + z3 + 4
-            sage: norm = phi.frobenius_norm()
-            sage: norm
-            (5*z3^2 + 2*z3)*T^2 + (4*z3^2 + 3*z3)*T + 5*z3^2 + 2*z3
+            sage: phi.frobenius_charpoly(algorithm="NotImplemented")
+            Traceback (most recent call last):
+            NotImplementedError: Algorithm "NotImplemented" not implemented
 
-        ::
+        ALGORITHM:
 
-            sage: n = 2  # Degree of the base field over Fq
-            sage: trace.degree() <= n/2
-            True
-            sage: norm.degree() == n
-            True
+            This wrapper dispatches the characteristic polynomial
+            computation to the method implementing the corresponding
+            algorithm. By default chooses the 'crystalline' algorithm.
+            See the corresponding docstring for each method for details.
+        """
+        algorithms = {'gekeler', 'crystalline'}
+        if algorithm in algorithms:
+            return getattr(self, \
+                    f'_frobenius_charpoly_{algorithm}')(var)
+        raise NotImplementedError(f'Algorithm \"{algorithm}\" not implemented')
 
-        ALGORITHM:
+    def _frobenius_charpoly_crystalline(self, var='X'):
+        r"""
+        Method to compute the characteristic polynomial of the
+        Frobenius endomorphism using Crystalline cohomology.
+        Currently only works for Drinfeld modules defined over
+        Fq[T], but otherwise does not impose any other constraints,
+        including on the rank, minimal polynomial, or that the Drinfeld
+        module is defined over the prime field.
+
+        INPUT:
+
+        - ``var`` (default: ``'X'``) -- the name of the second variable
+
+        EXAMPLES::
+
+            sage: Fq = GF(25)
+            sage: A.<T> = Fq[]
+            sage: K.<z> = Fq.extension(8)
+            sage: phi = DrinfeldModule(A, [z, 4, 1, z, z+1, 2, z+2, 1, 1, 3, 1])
+            sage: phi._frobenius_charpoly_crystalline()
+            X^10 + X^9 + (3*T + z2 + 1)*X^8 + (4*T^2 + z2*T + 2*z2 + 1)*X^7 + (4*T^3 + (z2 + 2)*T^2 + (4*z2 + 2)*T + 4)*X^6 + (3*T^4 + T^3 + 3*z2*T + 3*z2 + 3)*X^5 + ((4*z2 + 2)*T^4 + (3*z2 + 4)*T^3 + 3*T^2 + (2*z2 + 4)*T + 4*z2 + 1)*X^4 + (3*T^5 + (2*z2 + 3)*T^4 + 4*T^3 + (2*z2 + 2)*T^2 + (z2 + 3)*T + 4*z2)*X^3 + (3*T^6 + (3*z2 + 2)*T^5 + (4*z2 + 1)*T^4 + z2*T^3 + (4*z2 + 4)*T^2 + 4*z2*T)*X^2 + (2*T^7 + 3*T^6 + 2*z2*T^5 + (2*z2 + 3)*T^4 + (4*z2 + 3)*T^3 + (z2 + 2)*T^2 + (z2 + 4)*T + 2*z2 + 2)*X + T^8 + (4*z2 + 3)*T^6 + (4*z2 + 4)*T^4 + 4*z2*T^2 + (z2 + 2)*T + z2
```

Comment:
IMO, the complete output is not really meaningful; maybe, you should cut it using Ellipsis (`...`);